File: Hash/383.赎金信.cpp
/*
 * @lc app=leetcode.cn id=383 lang=cpp
 *
 * [383] 赎金信
 */

// @lc code=start
class Solution
{
public:
    bool canConstruct(string ransomNote, string magazine)
    {
        unordered_map<char, int> ransomMap;
        // map记录单个字符出现的次数
        for (char w : ransomNote)
        {
            ransomMap[w]++;
        }

        // 单个字符再次出现次数-1
        for (char w : magazine)
        {
            if (ransomMap.find(w) != ransomMap.end())
            {
                ransomMap[w]--;
            }
        }

        // 最后map依然处在次数大于1的字符，说明没法满足
        for (auto iter = ransomMap.begin(); iter != ransomMap.end(); ++iter)
        {
            if (iter->second > 0)
            {
                return false;
            }
        }

        return true;
    }
};
// @lc code=end

File: Hash/383.赎金信.ts
/*
 * @lc app=leetcode.cn id=383 lang=typescript
 *
 * [383] 赎金信
 */

// @lc code=start
function canConstruct(ransomNote: string, magazine: string): boolean {
  let result = {}

  for (let letter of ransomNote) {
    if (result[letter]) {
      result[letter] += 1
    } else {
      result[letter] = 1
    }
  }

  for (let letter of magazine) {
    if (result[letter]) {
      result[letter] -= 1
    }
  }

  for (let key in result) {
    if (result[key] > 0) {
      return false
    }
  }

  return true
};
// @lc code=end

